blocks_vertical/procedures.js

```diff
@@ -312,91 +345,208 @@ Blockly.ScratchBlocks.ProcedureUtils.attachArgumentReporter_ = function(
  */
-Blockly.ScratchBlocks.ProcedureUtils.updateDisplay_ = function() {
-  var wasRendered = this.rendered;
-  this.rendered = false;
+Blockly.ScratchBlocks.ProcedureUtils.attachShadowDeclarationRoot_ = function(
+    input, inputType, displayName) {
+  if (inputType == 'n' || inputType == 's') {
+    var newBlock = this.workspace.newBlock('text');
+  } else {
+    var newBlock = this.workspace.newBlock('boolean_textinput');
+  }
+  newBlock.setFieldValue(displayName, 'TEXT');
+  newBlock.setShadow(true);
+  if (!this.isInsertionMarker()) {
+    newBlock.initSvg();
+    newBlock.render(false);
+  }
+  newBlock.outputConnection.connect(input.connection);
+};
 
-  if (this.paramMap_) {
-    var connectionMap = this.disconnectOldBlocks_();
-    this.removeAllInputs_();
+/**
+ * Update the serializable information on the block based on the existing inputs
+ * and their text.
+ */
+Blockly.ScratchBlocks.ProcedureUtils.updateProcCodeDeclarationRoot_ = function() {
+  this.procCode_ = '';
+  this.displayNames_ = [];
+  this.argumentIds_ = [];
+  for (var i = 0; i < this.inputList.length; i++) {
+    if (i != 0) {
+      this.procCode_ += ' ';
+    }
+    var input = this.inputList[i];
+    if (input.type == Blockly.DUMMY_INPUT) {
+      this.procCode_ += input.fieldRow[0].getValue();
+    } else if (input.type == Blockly.INPUT_VALUE) {
+      var target = input.connection.targetBlock();
+      this.displayNames_.push(target.getFieldValue('TEXT'));
+      this.argumentIds_.push(input.name);
+      if (target.type == 'boolean_textinput') {
+        this.procCode_ += '%b';
+      } else {
+        this.procCode_ += '%s';
+      }
+    } else {
+      throw new Error(
+          'Unexpected input type on a procedure mutator root: ' + input.type);
+    }
   }
+};
 
-  this.createAllInputs_(connectionMap);
-  this.deleteShadows_(connectionMap);
+/**
+ * Externally-visible function to add a label field to the mutator root block.
+ * @public
+ */
+Blockly.ScratchBlocks.ProcedureUtils.addLabelExternal = function() {
+  this.procCode_ = this.procCode_ + ' label text';
+  this.updateDisplay_();
+};
 
-  // TODO: Maybe also bump all old non-shadow blocks explicitly.
+/**
+ * Externally-visible function to add a boolean field to the mutator root block.
+ * @public
+ */
+Blockly.ScratchBlocks.ProcedureUtils.addBooleanExternal = function() {
+  this.procCode_ = this.procCode_ + ' %b';
+  this.displayNames_.push('boolean');
+  this.argumentIds_.push(Blockly.utils.genUid());
+  this.argumentDefaults_.push('todo');
+  this.updateDisplay_();
+};
 
-  this.rendered = wasRendered;
-  if (wasRendered && !this.isInsertionMarker()) {
-    this.initSvg();
-    this.render();
-  }
+/**
+ * Externally-visible function to add a string/number field to the mutator root
+ * block.
+ * @public
+ */
+Blockly.ScratchBlocks.ProcedureUtils.addStringNumberExternal = function() {
+  this.procCode_ = this.procCode_ + ' %s';
+  this.displayNames_.push('string or number');
```

Comment:
does this need to be i18n'd?

Reply:
If we keep a string here then yes, it needs i18n.  I don't know whether the plan is to have text here by default.